Extracting same code of endpointslice and endpointslicemirror into a new shared EndpointSlice package

JornShen opened the pull request:
/kind cleanup

ref: https://github.com/kubernetes/kubernetes/pull/99756#issuecomment-790814861

This MR extract same code include:
error.go 
endpointSliceTracker.go 
endpointset.go



```release-note
NONE
```

aojea commented:
/cc

JornShen commented:
/cc @robscott @aojea

JornShen commented:
/cc @thockin @danwinship 
:)

aojea commented on pkg/controller/endpointslicemirroring/metrics.go:
```diff
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions and
 */
 
-package metrics
+package endpointslicemirroring
```
why removing this from its own folder?

aojea replied:
I don't know, I feel that is more readable in its own folder, all the metrics are prefixed.
I defer to @robscott for this

aojea commented on pkg/controller/endpointslicemirroring/reconciler.go:
```diff
@@ -197,14 +199,9 @@ func (r *reconciler) reconcileByPortMapping(
-		compareLabels := cloneAndRemoveKeys(existingSlices[0].Labels, discovery.LabelManagedBy, discovery.LabelServiceName)
-		// Return early if first slice matches desired endpoints, labels and annotations
+		// Return early if first slice matches desired endpoints.
 		totals = totalChanges(existingSlices[0], desiredSet)
-		if totals.added == 0 && totals.updated == 0 && totals.removed == 0 &&
-			apiequality.Semantic.DeepEqual(endpoints.Labels, compareLabels) &&
-			apiequality.Semantic.DeepEqual(compareAnnotations, existingSlices[0].Annotations) {
+		if totals.added == 0 && totals.updated == 0 && totals.removed == 0 {
```
you are modifying the behaviour, the goal is to consolidate the code from both controllers into one place, the end results should not modify any behaviour
maybe my comment about using previous PR as guidance was misleading?

JornShen replied:
rebased.

aojea commented on pkg/controller/util/endpointslice/cache.go:
```diff
@@ -106,21 +110,27 @@ func (c *Cache) UpdateEndpointPortCache(endpointsNN types.NamespacedName, epCach
 }
 
 // DeleteEndpoints removes references to an Endpoints resource from the global
 // cache and updates the corresponding metrics.
-func (c *Cache) DeleteEndpoints(endpointsNN types.NamespacedName) {
+func (c *Cache) DeleteEndpoints(portNN types.NamespacedName) (epExist bool, actual, desired, numEndpoints int) {
 	c.lock.Lock()
 	defer c.lock.Unlock()
```
This change is hard to understand without context, maybe you should use different commits for different changes, at least I´m having problems to follow the logic in this metrics cache changes

aojea commented on pkg/controller/endpointslice/utils.go:
```diff
@@ -394,3 +394,16 @@ func hintsEnabled(annotations map[string]string) bool {
+
+// managedByChanged returns true if one of the provided EndpointSlices is
+// managed by the EndpointSlice controller while the other is not.
+func managedByChanged(endpointSlice1, endpointSlice2 *discovery.EndpointSlice) bool {
+	return managedByController(endpointSlice1) != managedByController(endpointSlice2)
+}
+
+// managedByController returns true if the controller of the provided
+// EndpointSlices is the EndpointSlice controller.
+func managedByController(endpointSlice *discovery.EndpointSlice) bool {
```
This seems to be reused in at least 2 more places, is not possible to export and consolidate them?

robscott replied:
I think this is actually the most straightforward given the reliance on the local `controllerName` var. Although we could extract this in the future, I'm fine with keeping it local for now.

aojea commented on pkg/controller/endpointslice/metrics/cache.go:
```diff
@@ -1,161 +0,0 @@
-// with ideal endpoint distribution.
-func numDesiredSlices(numEndpoints, maxPerSlice int) int {
-	if numEndpoints <= maxPerSlice {
-		return 1
```
I think because otherwise the division will give you a decimal number lower than 1, and you have to cast to an int.

dashpole commented:
/assign @dashpole

dashpole commented:
It looks like this isn't modifying any instrumentation code, so it doesn't need an instrumentation reviewer.  If that isn't correct, please reassign me

/remove-sig instrumentation
/unassign @dashpole

aojea commented:
> It looks like this isn't modifying any instrumentation code, so it doesn't need an instrumentation reviewer. If that isn't correct, please reassign me
> 
> /remove-sig instrumentation
> /unassign @dashpole

this is a refactor for endpointslices, that has some metrics associated, I think that with Rob Scott will be enough
/assign @robscott

robscott commented:
Hey @JornShen, thanks for taking this on! I think it will be very helpful to have a shared endpointslice util package for code shared between these controllers. I think that should likely include everything related to EndpointSliceTracker, EndpointSet, and maybe the StaleCacheError. Although there's certainly shared metrics code, it's harder to translate that as it's not an exact copy like the rest of these components. I'd rather punt on that for later. 

I'm a bit concerned with how difficult this could make applying cherry picks to releases that did not have this shared util package. Not sure on the specifics here, but if anyone knows of anything we can do to make that process easier, it would be very appreciated. I know of at least one more EndpointSliceTracker change I'd like to make + cherry pick back. I'm hesitant to merge this in before that change, just for ease of cherry picking, but let me know if I'm missing something that would make that easier.

aojea commented:
> Hey @JornShen, thanks for taking this on! I think it will be very helpful to have a shared endpointslice util package for code shared between these controllers. I think that should likely include everything related to EndpointSliceTracker, EndpointSet, and maybe the StaleCacheError. Although there's certainly shared metrics code, it's harder to translate that as it's not an exact copy like the rest of these components. I'd rather punt on that for later.
> 
> I'm a bit concerned with how difficult this could make applying cherry picks to releases that did not have this shared util package. Not sure on the specifics here, but if anyone knows of anything we can do to make that process easier, it would be very appreciated. I know of at least one more EndpointSliceTracker change I'd like to make + cherry pick back. I'm hesitant to merge this in before that change, just for ease of cherry picking, but let me know if I'm missing something that would make that easier.

I think that we should make this change once we are confident with the status, definitely not in this release, but maybe 1.22?

JornShen commented:
> > Hey @JornShen, thanks for taking this on! I think it will be very helpful to have a shared endpointslice util package for code shared between these controllers. I think that should likely include everything related to EndpointSliceTracker, EndpointSet, and maybe the StaleCacheError. Although there's certainly shared metrics code, it's harder to translate that as it's not an exact copy like the rest of these components. I'd rather punt on that for later.
> > I'm a bit concerned with how difficult this could make applying cherry picks to releases that did not have this shared util package. Not sure on the specifics here, but if anyone knows of anything we can do to make that process easier, it would be very appreciated. I know of at least one more EndpointSliceTracker change I'd like to make + cherry pick back. I'm hesitant to merge this in before that change, just for ease of cherry picking, but let me know if I'm missing something that would make that easier.
> 
> I think that we should make this change once we are confident with the status, definitely not in this release, but maybe 1.22?

@aojea @robscott 

merging cache.go to shared code is my attempt. But It seems that looks not very perfect, change too much place :(
And I think we can wait for the time when these codes become more stable to make this change.  I dont mind that we put this refactor off the next version or further version. :)

JornShen committed:
extract same code of es and esm to pkg

migrate files:
endpointset.go
endpointslice_tracker.go
endpointslice_tracker_test.go
errors.go

JornShen commented:
/cc @aojea @robscott

JornShen commented:
ping @robscott

robscott reviewed (commented):
Thanks @JornShen! I think this reaches a good balance of deduplicating code while also not going too far. The components you've moved into a util package really are identical, and I think the resulting change looks good and will make further changes to these components easier to handle. Unfortunately cherry picking changes to these components will be a bit more difficult for a couple releases, but I think with the latest round of bug fixes already picked back, this has hopefully reached some level of stability.

/lgtm
/approve

Also assigning @thockin for the pkg/controller bits I can't approve.
/assign @thockin

aojea commented:
/assign @thockin 
for final approval

thockin commented on pkg/controller/endpointslice/endpointslice_controller.go:
```diff
@@ -141,7 +142,7 @@ func NewController(podInformer coreinformers.PodInformer,
 	c.endpointSliceLister = endpointSliceInformer.Lister()
 	c.endpointSlicesSynced = endpointSliceInformer.Informer().HasSynced
-	c.endpointSliceTracker = newEndpointSliceTracker()
+	c.endpointSliceTracker = endpointsliceutil.NewEndpointSliceTracker()
```
I see you using the new one, but not removing the old one?

JornShen replied:
@thockin  
I change old function name for public scope since we have moved the file `endpointslice_tracker.go` to new place. 
And I modfiy all the `newEndpointSliceTracker` function calling place

thockin commented on pkg/controller/util/endpointslice/endpointset.go:
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2019 The Kubernetes Authors.
+Copyright 2021 The Kubernetes Authors.
```
don't update copyright on file moves

JornShen replied:
done.

JornShen commented:
ping @thockin 
PTAL  :)

JornShen commented:
ping @thockin

JornShen commented:
@thockin @robscott Do you consider merging it in this version ?

aojea commented:
/lgtm

thockin commented:
/approve
